docs: add environment cloning and how-to guide (#2249)
Unleash/unleash: website/docs/how-to/how-to-clone-environment.mdx

---
title: How to clone an environment
---

:::info availability

> Environment cloning is planned to be released in **Unleash 4.18**.

Author: Unsure about this, just a guess. We might want to include it right away in 4.17.

Reviewer: Yeah, nice suggestion, we can make it more narrow when we know better

:::

> Environment cloning is a tool to enable an admin to clone an environment, including all feature toggles and strategies, into a new environment. See [_Cloning environments_](../user_guide/environments#cloning-environments) for more information.

Reviewer: - It's exactly the same thing but when I read "cloning is a tool to clone" it feels a bit redundant
- We're not technically cloning toggles because you can't do that, only their enabled state, I think not being clear about that has the potential to be very confusing

```suggestion
Environment cloning is a tool to enable an admin to duplicate an existing environment, including all feature toggles strategies and their state. See [_Cloning environments_](../user_guide/environments#cloning-environments) for more information.
```

Author: Thanks, much clearer 👍

## Step 1: Navigate to the environments page {#step-1}

Navigate to the _Environments_ page in the admin UI (available at the URL `/environments`). Use the navigation menu item "Configure" and select "Environments".

![The admin UI navigation "Configure" submenu with the Environments item highlighted.](/img/clone-environment-1.png)

## Step 2: Select an environment to clone {#step-2}

Select an environment to clone. On the right side, open the actions submenu and select "Clone".

![The "production" environment actions submenu with the Clone option highlighted.](/img/clone-environment-2.png)

## Step 3: Fill in the clone environment form {#step-3}

Give your new environment a name. The name must be unique and cannot be the same as the original environment. The name is pre-filled with a suggestion, but you can change it to whatever you like.

> Select an environment type, and which projects to clone the environment configuration in. You also have the option to keep the users permission for the new environment.

Reviewer 2: Hmm, I wonder if this can be clarified a bit further 🤔

What happens if they don't keep permissions?
```suggestion
Select an environment type and select which projects should have their environment configuration cloned. Choose whether to keep the existing user permissions for the new environment or ... [what happens here?]
```

Reviewer: Good point, you get a fresh environment with no RBAC overrides for create/update/delete strategy or turn on/off a toggle in that environment. Effectively only an admin or editor of the project can fiddle with the strategies/enable state for that environment

![The clone environment form filled with some example data, and the Clone environment button highlighted at the bottom.](/img/clone-environment-3.png)

## Step 4 (optional): Generate a new API token {#step-4}

> On the form, you also have the option to generate an API token right away. If you choose to do so, you will need to select the projects enabled for the new token. Once you submit the form a new server-side SDK (CLIENT) token will be generated for you.

Reviewer 2: If this is the same form, I think it should be part of the same step as before (step 3).

It could maybe be something like:
```suggestion
You can optionally generate an API token for the new environment right away. Select which projects the token should have access to, and the token will be generated when you submit the form.
```
